core/src/impl/Kokkos_ViewMapping.hpp

```diff
@@ -2994,7 +2994,14 @@ struct ViewValueFunctor<DeviceType, ValueType, true /* is_scalar */> {
     }
   }
 
-  void destroy_shared_allocation() {}
+  void destroy_shared_allocation() {
+    // TODO: Aren't views supposed to fence before being deleted? Fence all
+    // instances or only one? space is not necessarily the only one using the
+    // view. Does that happen elsewhere? TestScatterView fails with the
+    // asynchronous HPX backend unless there's a fence (e.g.) here. Are all
+    // internal views in scatter views reference counted? Hidden assumptions?
+    space.fence();
```

Comment:
This is for calling the destructor of the underlying types (that we only assume to be necessary for array types), not for freeing the memory which we assume to be synchronous (in `MemorySpace::deallocate`).

Reply:
I think I was mixing up the behaviour of the Cuda memory space and the host memory space. The Cuda space does synchronize in view destructors, right (at least when the ref count goes to zero)? And the host space does not?

You can ignore the fence here. What happens is that this block triggers a segfault with the asynchronous HPX backend: https://github.com/kokkos/kokkos/blob/e4831441fb213bd29485327b02cf1b54149a58a7/containers/unit_tests/TestScatterView.hpp#L446-L461. `create_scatter_view` spawns an asynchronous `parallel_for` with `ResetDuplicates` which, for big `n` does not complete before the end of the scope. Since there's no fence the parallel for loop segfaults trying to access views that have already been freed. I don't know if it's preferable to have a fence in the test, scatter view constructors, destructor, or when freeing host memory.

Reply:
Any comments on this? I've added a fence over here: https://github.com/kokkos/kokkos/blob/2649edcd57546bf0b5846d35d21b8bbe47e10af9/containers/unit_tests/TestScatterView.hpp#L472 makes the test pass. Would someplace else be more appropriate?

Reply:
I think we just need a fence in `HostSpace::impl_deallocate()` fencing all the execution spaces that could access host memory.

Reply:
Do you already have functionality to do that generically? `Kokkos::fence()` is too heavy and `Kokkos::DefaultHostExecutionSpace().fence()` is too light. I presume this is something that `ExecSpaceManager` would have to take care of?

Reply:
No, we don't have that functionality. I think we want something like `Kokkos::fence_host` that only fences the execution spaces for which
```
SpaceAccesibility<ExecutionSpaceType, Kokkos::HostSpace>::accessible
```
is `true`.

Reply:
We talked about it in the meeting and we would just go with a global fence, i.e. `Kokkos::fence()`.

Reply:
Would you be willing to just address that in this pull request?

Reply:
Oh, it looks like you already have it at https://github.com/kokkos/kokkos/pull/4241/files#diff-c4732ec78317e1eb713a2e4eb3397fc76b36b0ea64b84c566ca0f1a29f320e76R311.

Reply:
I pushed the change, but did not test it, hence I didn't comment yet... I suppose you guys are ok with the tradeoffs of the fence at the moment given that you've discussed it? I could imagine it even affecting users who aren't using the HPX backend (e.g. expecting being able to do work on the CPU while things are running asynchronously on the GPU). For us the fence is fine (though a host-only fence would obviously be preferable in the long run).

Reply:
Just to report back, this does fix the view mapping test, but hangs in `impl_shared_alloc`. My guess is that it fences inside a parallel region...